### ISSUE TITLE
[Workflow] Some workflow proposals, including ZenHub

### ISSUE THREAD

**Issue by electricjones** [spider/spider#124]
I have finished up a couple other projects, and have decided to put all my other personal things on hold to get Spider going. I really enjoy this project and I feel the time is right. Since I am going to be working more on it (and hopefully recruiting others), there are some workflow things I wanted to propose.

(Check off after decided to implement)
- [x] Move to docker, as mentioned in #120 
- [x] Introduce a basic CLI tool, as mentioned in #123 
- [x] Use [ZenHub](https://www.zenhub.io/) for project management. This would of course be optional, but it would be nice for the core team to be on the same page.
- [x] Organize and use labels more intentionally. Discussion needed.
- [x] Prefix Issues (and PRs) with a component tag [Workflow], [Models], [Builders], etc. Discussion needed to identify components.
- [x] Have a standard commit message format that includes the issue number (if applicable)

All of this would require updating documentation.

After the next release, I may also look into starting a simple blog/marketing site through github pages. After this next release, we should have something actually usable.

**Comment by smolinari** [comment]
I'm for Zenhub too. I've been looking for a reason to try it.

Labels, not sure how I can offer much to the discussion. 

Documentation, I can definitely help there. 

Who are you looking to recruit?

Scott

**Comment by electricjones** [comment]
No one in particular. Once we get a working release out, I thought about reaching out on various neo, gremlin, and orient boards

**Comment by dmill-bz** [comment]
Hey guys, 

Just a heads up I'm totally drowning in work atm, sorry I haven't been as active lately. 
@chrismichaels84 I've been looking for php devs with good graph knowledge to help out on some of my projects (payed) and I was wondering if you may be up for it. It wouldn't start right away but I'm basically building a pool of people I can call up for work. Is this something you could be interested in? I found your phone number via your online resume, if interested when's a good time for a call?

**Comment by electricjones** [comment]
I don't think now would be a good time. I've also signed some non-competes with my company. Not sure what qualifies exactly, but my summer is booked pretty solid.

**Comment by dmill-bz** [comment]
Ok! Like I said, it wouldn't be for right away, I actually wasn't looking at anything this summer (maybe fall or winter 2016). But let me know if it's too much of a headache in regards to your non-competes. 
Cheers.

**Comment by smolinari** [comment]
Does that mean Spider will stay on the back burner for a good while? :sob: 

Scott

**Comment by electricjones** [comment]
Alright, we can pick up this conversation closer to the fall. I'll check on my non-competes, but I think they are more geared toward real estate than other development. And, @smolinari, I have spider planned for this summer.

**Comment by electricjones** [comment]
Yeah, it's actually on the scope for something my team wants to do, and will likely be the core of a non profit project I'm wanting to work on starting in the fall. So, I have real plans to get a working copy up by summer's end.

For me, the real hurdle is the motivation to learn gremlin ;)

**Comment by electricjones** [comment]
My proposals for some of the workflow suggestions: (this post to be added to documentation)

## Zenhub

I have setup the following columns to the board: 
1. Backlog - For new items that have not been categorized yet
2. Groomed - Items that have been groomed, discussed, and probably assigned and estimated. Ready for work
3. Discussion - Items that need discussion before real work begins (includes working-draft proposals)
4. In Progress - Items that have been claimed by one or more people and active work is being done
5. Waiting - Waiting on a dependency (NOT to be used for "waiting on a person to get to it")
6. Code Review - PR finished and ready for review
7. Long Running - A handful of issues that never really close
8. Closed - All done skippy ;)

Zenhub things to consider
- It would be nice to use estimates and log work
- Assign issues to people, can be multiple assignment (now)

## Workflow Labels and Components

Let's keep the github labels for workflow items and compontent tags [Models] for components.

Component examples:
- [Builders] - for Command Builders
- [Tests] - For test-specific tools (or individual tests if that is ALL the issue is covering)
- [General Business] - For workflow, ideas, and non-code things
- [App] - Command line apps
- [Specific Database/Tool] like docker, neo4j, etc
- [ORM] when that comes around
- [Factory] - top level `Spider` object
- [Dev] - For development environment (docker, ci, etc)
- [Docs] - duh

Any other obvious ones I'm missing?

Labels, then, are more about workflow, position, and communication. Used to describe the issue.

I think these are pretty up-to-date. Examples: Bug, proposal, discussion, long running, urgent, help wanted, etc

Thoughts?

**Comment by electricjones** [comment]
As far as the commit messages, I'd suggest the following guidelines:
1. The first line should always be 50 characters or less and that it should be followed by a blank line.
1. Be in the present tense (fixes something, does something else, closed the third thing)
1. Link to the issue (using the #<id> format)
1. Fine to [close issue via commit message](https://help.github.com/articles/closing-issues-via-commit-messages/), but only when it's actually closed
1. Should be in the following format `#123: Some message here`

Pull requests can handle multiple issues, and should have a clear and descriptive title. There should also be a reference list in the opening comment with the issues the pr intends to address.

Issues and pull requests should use labels and component tags. Commit messages should reference the issue number, but not use labels or tags.

### Examples

#### Issues
1. Add some cool feature (does not have a component tag bc its global) <help wanted> <proposal>
2. [ORM] Returns wrong type when invoked via magic method <urgent> <bug>
3. [Docs] Add #124 to contributing documentation

#### Pull Requests

[Builders] Adds query saving and reuse <enhancement>
Includes issues #x, #y, and #z

#### Commit Messages
1. #x: initializes query saving skeleton
2. #y: Removes some bug from Cypher query processor
3. #z: Closes #z by removing dead code
